build_tools/third_party/cuda/CMakeLists.txt

```diff
@@ -4,80 +4,183 @@
-    --label "${_VERSION}"
-    --product "${_PRODUCT}"
-    --os "${_OS}"
-    --arch "${_ARCH}"
-    --component "${COMPONENT}"
-    --output "${_DOWNLOAD_DIR}")
-endforeach()
-
-if(NOT EXISTS "${SRC_DIR}")
-  message(FATAL_ERROR "Download did not produce expected source dir: ${SRC_DIR}")
-  return()
+if(EXISTS "${IREE_CUDA_LIBDEVICE_PATH}")
+  message(STATUS "Using CUDA libdevice: ${IREE_CUDA_LIBDEVICE_PATH}")
+else()
+  message(SEND_ERROR "Cannot find CUDA libdevice file (${IREE_CUDA_LIBDEVICE_PATH}). Either configure your CUDA SDK such that it can be found or specify explicitly via -DIREE_CUDA_LIBDEVICE_PATH=/path/to/libdevice.10.bc")
 endif()
 
-foreach(REL_PATH ${_RETAIN_PATHS})
-  set(SRC_FILE "${SRC_DIR}/${REL_PATH}")
-  set(TARGET_FILE "${_TARGET_DIR}/${REL_PATH}")
-  message(STATUS "Copy ${SRC_FILE} -> ${TARGET_FILE}")
-  # file(COPY) expects a destination directory.
-  get_filename_component(TARGET_DIR "${TARGET_FILE}" DIRECTORY)
-  file(COPY ${SRC_FILE} DESTINATION ${TARGET_DIR})
-endforeach()
-
-# Delete tmp directory.
-file(REMOVE_RECURSE ${_DOWNLOAD_DIR})
-
-# Set vars for downloaded cuda deps
-set(IREE_CUDA_DOWNLOAD_LIBDEVICE_PATH
-    "${_TARGET_DIR}/nvvm/libdevice/libdevice.10.bc" PARENT_SCOPE)
-set(IREE_CUDA_DOWNLOAD_INCLUDE_PATH "${_TARGET_DIR}/include" PARENT_SCOPE)
+# Locate runtime components.
+if(CUDAToolkit_FOUND)
+  message(STATUS "Using CUDA INCLUDE_DIRS from found SDK: ${CUDAToolkit_INCLUDE_DIRS}")
+elseif(CUDAToolkit_ROOT)
+  # See note above about computers being hard.
+  # We make minimal use of CUDA for the runtime and really just need cuda.h
+  # presently. So let's make a guess at that.
+  set(CUDAToolkit_INCLUDE_DIRS "${CUDAToolkit_ROOT}/include")
+  if(EXISTS "${CUDAToolkit_INCLUDE_DIRS}/cuda.h")
+    message(STATUS "Using CUDA INCLUDE_DIRS from CUDAToolkit_ROOT: ${CUDAToolkit_INCLUDE_DIRS}")
+  else()
+    message(SEND_ERROR "Using explicitly specified CUDAToolkit_ROOT, could not find cuda.h at: ${CUDAToolkit_INCLUDE_DIRS}")
+  endif()
+elseif(IREE_CUDA_DOWNLOAD_INCLUDE_PATH)
+  message(STATUS "Using downloaded CUDA includes")
+  set(CUDAToolkit_INCLUDE_DIRS "${IREE_CUDA_DOWNLOAD_INCLUDE_PATH}")
+else()
+  message(SEND_ERROR "Cannot build IREE runtime CUDA components (-DIREE_HAL_DRIVER_CUDA=ON) because a CUDA SDK was not found. Consult CMake docs for your version: https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html")
+endif()
+
+################################################################################
+# Targets that IREE depends on which encapsulate access to CUDA.
+################################################################################
+
+iree_c_embed_data(
+  PACKAGE
+    iree_cuda
+  NAME
+    libdevice_embedded
+  SRCS
+    "${IREE_CUDA_LIBDEVICE_PATH}"
+  C_FILE_OUTPUT
+    "iree_cuda/libdevice_embedded.c"
+  H_FILE_OUTPUT
+    "iree_cuda/libdevice_embedded.h"
+  INCLUDES
+    # Allows an include like "iree_cuda/cuda_libdevice.h"
```

Comment:
Do we want such an include? I don't see any such usages in the codebase

Reply:
Stale comment (renamed the file). Updated: see CUDATarget.cpp.

Reply:
Hmmm I still don't actually see any usage: https://cs.github.com/iree-org/iree?q=libdevice_embedded

Reply:
The header was renamed... it's there in this patch set. Promise :)